python/src/iceberg/avro/reader.py

```diff
@@ -0,0 +1,318 @@
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    ListType,
+    LongType,
+    MapType,
+    NestedField,
+    PrimitiveType,
+    StringType,
+    StructType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+)
+from iceberg.utils.singleton import Singleton
+
+
+@dataclass(frozen=True)
+class AvroStruct(StructProtocol):
+    _data: list[Any | StructProtocol] = field()
+
+    def set(self, pos: int, value: Any) -> None:
+        self._data[pos] = value
+
+    def get(self, pos: int) -> Any:
+        return self._data[pos]
+
+
+class Reader(metaclass=Singleton):
+    @abstractmethod
+    def read(self, decoder: BinaryDecoder) -> Any:
+        ...
+
+
+class NoneReader(Reader):
+    def read(self, _: BinaryDecoder) -> None:
+        return None
+
+
+class BooleanReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> bool:
+        return decoder.read_boolean()
+
+
+class IntegerReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> int:
+        return decoder.read_int()
+
+
+class LongReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> int:
+        return decoder.read_long()
+
+
+class FloatReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> float:
+        return decoder.read_float()
+
+
+class DoubleReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> float:
+        return decoder.read_double()
+
+
+class DateReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> date:
+        return decoder.read_date_from_int()
+
+
+class TimeReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> time:
+        return decoder.read_time_micros_from_long()
+
+
+class TimestampReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> datetime:
+        return decoder.read_timestamp_micros_from_long()
+
+
+class TimestamptzReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> datetime:
+        return decoder.read_timestamp_micros_from_long()
+
+
+class StringReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> str:
+        return decoder.read_utf8()
+
+
+class UUIDReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> UUID:
+        return UUID(decoder.read_utf8())
+
+
+@dataclass(frozen=True)
+class FixedReader(Reader):
+    length: int = field()
+
+    def read(self, decoder: BinaryDecoder) -> bytes:
+        return decoder.read(self.length)
+
+
+class BinaryReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> bytes:
+        return decoder.read_bytes()
+
+
+@dataclass(frozen=True)
+class DecimalReader(Reader):
+    precision: int = field()
+    scale: int = field()
+
+    def read(self, decoder: BinaryDecoder) -> Decimal:
+        return decoder.read_decimal_from_bytes(self.precision, self.scale)
+
+
+@dataclass(frozen=True)
+class OptionReader(Reader):
+    option: Reader = field()
+
+    def read(self, decoder: BinaryDecoder) -> Any | None:
+        # For the Iceberg spec it is required to set the default value to null
```

Comment:
This comment is true, but for a long time there was nothing in Avro that actually enforced this so I've definitely seen out of order schemas. I would opt to figure out which option is non-null and test for it directly:

```python
class OptionReader(Reader):
    option_reader: Reader = field()
    non_null_option: int = field()

    def read(self, decoder: BinaryDecoder) -> Any | None:
        if decoder.read_int() == non_null_option:
            return self.option_reader.read(decoder)
        return None
```

Reply:
![oof](https://user-images.githubusercontent.com/1134248/173377856-6eb8c829-d6ab-43b6-819d-4a8369f07e45.gif)

We lose this information when we go to the Iceberg schema, since we don't keep the position there. This would either require storing this information in the Iceberg type, which I don't really like. Or build the reader from the Avro schema directly.

Reply:
Ah, I see. Let me think about this more, but it should be fine for now.

Reply:
I'm not able to judge how often this happens in practice. Once we run into it, we can change it